### PR TITLE
Add post:JS_what_is_functor

### PR DESCRIPTION
**Pull request by yiyu0x**

### CODE REVIEW

File: source/_posts/javascript-functor.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+title: 每天都在 arr.map()，你聽過 functor 嗎？
+date: 2019-07-18 15:59:38
+tags:
+    - javascript
+    - functor
+---
+
+## 前言
+
+大家都知道程式語言分為兩種，分為 `functional programming` 以及 `object oriented programming`。然而看過不少人有一種迷思，認為**物件導向的語言一定比較好**，無論是維護性，或是物件導向的種種特性，都可以將抽象化的概念更加發揮至淋漓盡致。
```
Comment (aszx87410) [inline source/_posts/javascript-functor.md]: 我覺得這邊的觀念有點錯誤，我不覺得可以靠 functional 或是 OO 把程式語言來分類，因為 functional 與 OO 不是一個互斥的情形。再者，functional programming 指的不是程式語言，是一種 paradigm（一種覺得這詞翻中文有點怪XD）。

這邊引用一下 wiki 的定義：

> In computer science, functional programming is a programming paradigm—a style of building the structure and elements of computer programs—that treats computation as the evaluation of mathematical functions and avoids changing-state and mutable data

與其說某個語言是「functional programming language」，我認為更精確的說法是：「這個程式語言支援 functional programming」，這兩者是不同的。

就像 wiki 所寫的，`programming languages that support functional programming`：

> However, programming languages that support functional programming have been used in industry, including Common Lisp, Scheme,[3][4][5][6] Clojure, Wolfram Language[7][8], Racket,[9] Erlang,[10][11][12] OCaml,[13][14] Haskell,[15][16] and F#.[17][18] JavaScript, one of the world's most widely distributed languages,

底下也有提到說有些不是針對 FP 來設計的程式語言也可以 FP：

> Programming in a functional style can be accomplished in languages that are not specifically designed for functional programming, such as with Perl[27], PHP[28], C++11, and Kotlin

綜合以上，我覺得 FP 跟 OO 只是一個 paradigm，硬要跟程式語言綁在一起似乎不是個好選擇？

Comment (aszx87410) [reply source/_posts/javascript-functor.md]: https://en.wikipedia.org/wiki/Functional_programming

**Comment by aszx87410** [comment]
然後我覺得這篇文章的 scope 有點太小，只提了 functor，其他很多東西都只有點到而已。有點像是我今天只講 XMLHttpRequest 裡面 onload 這個方法，其他都稍微點一下。

我自己是會比較期待更完整的主題啦，例如說要講 FP 就從頭開始講然後帶到 functor，裡面的很多名詞也可以順便一起做補充，會完整很多。

不知道 @ArvinH 跟 @kdchang 怎麼看？

File: source/_posts/javascript-functor.md (hunk 1 above)
Comment (yiyu0x) [reply source/_posts/javascript-functor.md]: 了解你的意思。我那句話想表達的是像 Java 這種先天支援物件導向的語言去跟 C 這種本身沒有辦法產生新物件的語言之間的差異。我用詞似乎不太妥當，我修改並且審視之後再重新提交。

**Comment by yiyu0x** [comment]
感謝 @aszx87410 的意見。我起初是認為可以分成好幾篇，如果覺得這樣內容太過空泛，也可以寫成一篇長文探討多一點 FP 的各種特性。

**Comment by arvinhv** [comment]
> 我自己是會比較期待更完整的主題啦，例如說要講 FP 就從頭開始講然後帶到 functor，裡面的很多名詞也可以順便一起做補充，會完整很多。

同意這點，希望能補充完整一點的長文：）

**Comment by kdchang** [comment]
Hi @yiyu0x,

Sorry for late reply. I think that it's a precious thing that a student wants to study and sharing FP knowledge, so I have invited my friends @wuct who is the FP expert as a reviewer. He will mentor you to optimize your posts and sharp your knowledge in FP. I think it's a golden opportunity, cheer up:)

BTW, we would gradually migrate to new writing submission mechanism and award system, maybe you can try it next time. thank you :)

Ref: https://www.coderbridge.com/articles/c63f676ec1ad4e1e8e8bd1f5abb1bcc4

**Pull request closed without merging**